### PR TITLE
[MRG+2] Switch testing to pytest completely

### PR DESCRIPTION
**Pull request by QuLogic**
I'm pre-opening this PR to verify that tests continue to build correctly. However, it still depends on #7935 and #7973.

This PR stops installing nose, and uses pytest everywhere. Note that I did not yet delete any redundant builds. I want to make sure everything works with the different methods first and will then see about pruning stuff.

### CODE REVIEW

**Comment by tacaswell** [comment]
attn @story645 I think I recognize this issue...

```
_________________________ TestUnitData.test_update_map _________________________
[gw0] linux2 -- Python 2.7.9 /home/travis/build/matplotlib/matplotlib/venv/bin/python
self = <matplotlib.tests.test_category.TestUnitData object at 0x7fffdfdb6910>
    def test_update_map(self):
        data = ['a', 'd']
        oseq = ['a', 'd']
        olocs = [0, 1]
    
        data_update = ['b', 'd', 'e', np.inf]
        useq = ['a', 'd', 'b', 'e', 'inf']
        ulocs = [0, 1, 2, 3, -2]
    
        unitdata = cat.UnitData(data)
        assert unitdata.seq == oseq
        assert unitdata.locs == olocs
    
        unitdata.update(data_update)
>       assert unitdata.seq == useq
E       assert [u'a', u'd', u'b', u'e', u'i'] == ['a', 'd', 'b', 'e', 'inf']
E         At index 4 diff: u'i' != u'inf'
E         Use -v to get the full diff
lib/matplotlib/tests/test_category.py:44: AssertionError
```

@QuLogic Might want to turn off the fast-fail while we debug this?

**Comment by story645** [comment]
What version of numpy is that test using? that looks like the fun joy where numpy pre like 1.8 truncates inf and nan to the length of the largest string in the array,  which in this case is 1.

**Comment by QuLogic** [comment]
It looks like NumPy 1.7.1.

Though I have a bigger question of why so few tests seem to be run compared to locally.

**Comment by NelleV** [comment]
Side note: that PR is going to conflict against #7970

File: tests.py [hunk 1]
```diff
@@ -4,10 +4,9 @@
 #
```
Comment (NelleV) [inline tests.py]: Can't we just remove this script?

File: lib/matplotlib/tests/test_ticker.py [hunk 2]
```diff
@@ -240,7 +237,7 @@ def _sub_labels(axis, subs=()):
     assert label_test == label_expected
 
 
-@cleanup(style='default')
+@pytest.mark.style('default')
```
Comment (NelleV) [inline lib/matplotlib/tests/test_ticker.py]: Why is the style classic being replaced by default?

File: .travis.yml [hunk 3]
```diff
@@ -155,20 +151,20 @@ script:
         rm -rf ~/.cache/matplotlib
       fi
       export MPL_REPO_DIR=$PWD  # needed for pep8-conformance test of the examples
+      # Workaround for pytest-xdist flaky colletion order
```
Comment (NelleV) [inline .travis.yml]: nitpick: There is a typo here: colletion -> collection

**Comment by NelleV** [comment]
@QuLogic We are setting the option to step testing after the first failure to speed up the tests.

File: lib/matplotlib/tests/test_ticker.py (hunk 2 above)
Comment (QuLogic) [inline lib/matplotlib/tests/test_ticker.py]: Both lines say default here.

File: tests.py (hunk 1 above)
Comment (QuLogic) [inline tests.py]: Not against it, but also not going to make that decision unilaterally.

File: lib/matplotlib/tests/test_ticker.py (hunk 2 above)
Comment (NelleV) [inline lib/matplotlib/tests/test_ticker.py]: Indeed, they do… I thought I had seen some classic being converted to default, but I guess we'll see this once the tests run on travis anyways.

File: tests.py (hunk 1 above)
Comment (tacaswell) [inline tests.py]: I strongly prefer this script stay and remain functional.

File: lib/matplotlib/tests/test_ticker.py (hunk 2 above)
Comment (QuLogic) [inline lib/matplotlib/tests/test_ticker.py]: There are a few that do; see #7973 for an explanation.

**Comment by QuLogic** [comment]
I've marked this `[WIP]` so that people focus on #7973 before this one; plus I have to figure out AppVeyor.

**Comment by QuLogic** [comment]
I believe this should be working now, once CI catches. It currently includes #8001 and #8002 as well as several bug fixes that were missed due to the CI not running everything:
 * Change `stacklevel` of warning in `Figure.gca` because it's triggered in two tests and on Python 2.7 it doesn't get triggered properly in the second test.
 * Skip broken categories tests temporarily (see #7988)
 * Fix broken `!=` comparison with `SubplotSpec`
 * Normalize paths when looking for fonts because Windows requires it.
 * Remove one last `xfail` that I missed in earlier PRs.

**Comment by tacaswell** [comment]
power-cycled as the code being run on travis (from reading the errors) was not matching the code in the PR. I saw something like this on one of my day-job projects yesterday too.  I think the web-hook is firing before the `+refs/PRNUMBER/merge` references are updating or something...

**Comment by QuLogic** [comment]
We're good to go here now.

File: .travis.yml [hunk 4]
```diff
@@ -225,7 +220,7 @@ after_success:
     else
       echo "Will only deploy docs build from matplotlib master branch"
     fi
-    if [[ $NOSE_ARGS =~ "--with-coverage" || $USE_PYTEST == true ]]; then
+    if [[ $USE_PYTEST == true ]]; then
```
Comment (tacaswell) [inline .travis.yml]: We should do this on all runs now?

Comment (QuLogic) [inline .travis.yml]: I was leaving that to #8003, but I see now it doesn't change this condition.

Comment (dopplershift) [inline .travis.yml]: Go ahead and turn it on--I'll rebase on it.

**Comment by dstansby** [comment]
The change to pytest will need documenting, but I've opened up a new issue for that at #8015.

**Comment by tacaswell** [comment]
1: `======= 6470 passed, 10 skipped, 6 xfailed, 8 xpassed in 644.13 seconds ========` 6502
2: `======= 6477 passed, 12 skipped, 2 xfailed, 3 xpassed in 609.06 seconds ========` 6497
3: `======== 6485 passed, 4 skipped, 2 xfailed, 3 xpassed in 611.43 seconds ========` 6497
4: docs
5: `======== 6485 passed, 4 skipped, 2 xfailed, 3 xpassed in 617.37 seconds ========` 6497
6: `======= 7259 passed, 12 skipped, 2 xfailed, 3 xpassed in 610.41 seconds ========` 7276
7: `====== 5256 passed, 1227 skipped, 10 xfailed, 1 xpassed in 573.52 seconds ======` 6494
8: `======= 6477 passed, 12 skipped, 2 xfailed, 3 xpassed in 579.11 seconds ========` 6494

There seems to be some slight descrepencies in the number of tests each run finds.

**Comment by dstansby** [comment]
Adding up those numbers I get 6494 for everything but (6) for which I get 7276.

**Comment by dstansby** [comment]
And I'm guessing that the extra checks on (6) are PEP8 checks.

**Comment by tacaswell** [comment]
Apparently I can not use a calculator :sheep:

**Pull request merged**